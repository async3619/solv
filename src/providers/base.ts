import fetch from "node-fetch";

import { Challenge, InputType } from "../utils/types";
import { CachedChallenge } from "../utils/retrieveChallenge";

export abstract class BaseProvider {
    public abstract checkUrl(url: string): boolean;
    public abstract retrieve(url: string): Promise<Challenge>;
    public abstract getId(url: string): string | number;

    protected constructor(public readonly needJs: boolean) {}

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public hydrateCache({ version, ...challenge }: CachedChallenge): Challenge {
        return {
            ...challenge,
            provider: this,
        };
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem !== "string" ? JSON.stringify(outputItem) : outputItem;
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public beforeExecute(sourceCode: string, input: InputType): string {
        return sourceCode;
    }

    public getName(): string {
        return this.constructor.name.replace("Provider", "");
    }

    protected async fetch(url: string): Promise<string> {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch (${response.status}) - ${response.statusText}`);
        }

        return response.text();
    }
}
