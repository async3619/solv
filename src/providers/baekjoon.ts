import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";
import { Challenge } from "../types";

export class BaekjoonProvider extends BaseProvider {
    private readonly urlRegex = /^https?:\/\/(www\.)?acmicpc\.net\/problem\/([0-9]*?)$/;

    public checkUrl(url: string) {
        return this.urlRegex.test(url);
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        const htmlCode = await this.fetch(url);
        const document = parseDOM(htmlCode);

        const titleDOM = document.querySelector("#problem_title");
        const descriptionDOM = document.querySelector("#problem_description");
        const inputDescriptionDOM = document.querySelector("#problem_input");
        const outputDescriptionDOM = document.querySelector("#problem_output");
        const ioItem = document.querySelectorAll(".sampledata");

        const ioText: [boolean, string][] = [];
        for (const item of ioItem) {
            ioText.push([item.id.includes("-input-"), item.innerHTML.trim()]);
        }

        if (
            !titleDOM?.textContent ||
            !descriptionDOM?.textContent ||
            !inputDescriptionDOM?.textContent ||
            !outputDescriptionDOM?.textContent
        ) {
            throw new Error("Failed to parse challenge information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            inputDescription: inputDescriptionDOM.textContent.trim(),
            outputDescription: outputDescriptionDOM.textContent.trim(),
            input: ioText.filter(([i]) => i).map(([, t]) => t),
            output: ioText.filter(([i]) => !i).map(([, t]) => t.replace(/\r\n/g, "\n")),
            provider: this,
            initialCode: `
function solution(input: string[]) {
    // ...
}

(() => {
    const readline = require("readline");

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const input: string[] = [];
    rl.on("line", line => {
        input.push(line);
    }).on("close", function () {
        solution(input);
        process.exit();
    });
})();
`.trim(),
        };
    }
}
