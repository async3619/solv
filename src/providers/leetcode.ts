import { assertParse, is } from "typia";
import { parse as parseDOM } from "node-html-parser";
import { Client, CombinedError, createClient } from "@urql/core";
import { stripHtml } from "string-strip-html";

import { BaseProvider } from "./base";

import { Challenge, InputType } from "../utils/types";
import { LeetCodeQuestionDocument } from "../queries/leetcode";
import type { LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables } from "../queries.data";
import { QuestionMetadata } from "./leetcode.types";

const URL_CHECK_REGEX = /(https?:\/\/)?leetcode\.com\/problems\/([a-z0-9-]+)\/?$/;

export class LeetCodeProvider extends BaseProvider {
    private readonly client: Client;

    public constructor() {
        super(false);

        this.client = createClient({
            url: "https://leetcode.com/graphql",
            requestPolicy: "network-only",
            fetchOptions: () => {
                return {
                    method: "GET",
                    headers: {
                        "content-type": "application/json",
                    },
                };
            },
        });
    }

    public checkUrl(url: string): boolean {
        return URL_CHECK_REGEX.test(url);
    }

    public getId(url: string): string | number {
        const match = URL_CHECK_REGEX.exec(url);
        if (!match) {
            throw new Error("Failed to parse challenge ID.");
        }

        return match[2];
    }

    public beforeExecute(sourceCode: string, input: InputType, challenge: Challenge): string {
        let entryPointName = "solution";
        if (challenge.initialCode) {
            const match = challenge.initialCode.match(/function ([a-zA-Z0-9]+)\(/);
            if (match) {
                entryPointName = match[1];
            }
        }

        return [sourceCode, `console.info(${entryPointName}(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const titleSlug = this.getId(url);
        if (!is<string>(titleSlug)) {
            throw new Error("Failed to parse challenge ID.");
        }

        try {
            const { error, data } = await this.client
                .query<LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables>(LeetCodeQuestionDocument, { titleSlug })
                .toPromise();

            if (error) {
                throw error;
            }

            if (!data?.question) {
                throw new Error("Failed to retrieve challenge data.");
            }

            const codeSnippet = data.question.codeSnippets.find(item => item.langSlug === "typescript");
            if (!codeSnippet) {
                throw new Error("This challenge does not support TypeScript.");
            }

            const document = parseDOM(data.question.content);
            const testCases = [...document.querySelectorAll("pre")].filter(item => {
                if (!item.parentNode) {
                    return false;
                }

                if (!item.previousElementSibling) {
                    return false;
                }

                const title = item.previousElementSibling.querySelector("strong.example");
                if (!title) {
                    return false;
                }

                return true;
            });

            const metadata = assertParse<QuestionMetadata>(data.question.metaData);
            const testCaseSummaries = testCases.map(item => stripHtml(item.textContent.trim()).result);
            const inputs: InputType[] = [];
            const outputs: string[] = [];
            for (const summary of testCaseSummaries) {
                const [input, output] = summary.split("\n").map(item => {
                    return item.trim().replace(/^[A-Za-z]*?: /, "");
                });

                const inputParameters = input.split(/, /).map(item => item.split(" = "));
                const data: InputType = [];
                for (const [key, value] of inputParameters) {
                    const index = metadata.params.findIndex(item => item.name === key);
                    if (index === -1) {
                        throw new Error(`Failed to find parameter "${key}".`);
                    }

                    data[index] = value;
                }

                inputs.push(data);
                outputs.push(output.replace("Output: ", ""));
            }

            return {
                id: data.question.questionId,
                title: data.question.questionTitle,
                description: stripHtml(data.question.content).result,
                provider: this,
                initialCode: codeSnippet.code,
                input: inputs,
                output: outputs,
            };
        } catch (e) {
            if (e instanceof CombinedError) {
                if (e.networkError) {
                    throw e.networkError;
                } else if (e.graphQLErrors && e.graphQLErrors.length > 0) {
                    throw e.graphQLErrors[0];
                }
            }

            throw e;
        }
    }
}
