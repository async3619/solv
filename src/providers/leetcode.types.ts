export interface QuestionMetadata {
    name: string;
    params: Array<{
        name: string;
        type: string;
    }>;
    return: {
        type: string;
    };
}
