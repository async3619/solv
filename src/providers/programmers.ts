import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";

import { findSiblingElement } from "../utils/findSiblingElement";
import { parseSampleTable } from "../utils/parseSampleTable";
import { Challenge } from "../utils/types";

export class ProgrammersProvider extends BaseProvider {
    public constructor() {
        super(false);
    }

    public checkUrl(url: string): boolean {
        return /(https?:\/\/)?(school\.)?programmers\.co\.kr\/learn\/courses\/([0-9]*?)\/lessons\/[0-9]([0-9]*?)$/.test(
            url,
        );
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem === "string" ? `"${outputItem}"` : JSON.stringify(outputItem);
    }

    public beforeExecute(sourceCode: string): string {
        return [sourceCode, `console.info(solution(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        targetUrl.searchParams.set("language", "javascript");

        const htmlCode = await this.fetch(targetUrl.toString());

        const document = parseDOM(htmlCode);
        const descriptionDOM = document.querySelector("#tour2 > div > div");
        const titleDOM = document.querySelector("#tab > li");
        if (!titleDOM?.textContent || !descriptionDOM?.textContent) {
            throw new Error("Failed to parse challenge information.");
        }

        const headers = document.querySelectorAll(".guide-section h5");
        const ioSampleHeaderRegex = /^입출력 ?예$/;
        const ioSampleHeaderDOM = headers.find(header => ioSampleHeaderRegex.test(header.textContent.trim()));
        if (!ioSampleHeaderDOM) {
            throw new Error("Failed to parse challenge input / output samples.");
        }

        const ioSampleTable = findSiblingElement(ioSampleHeaderDOM, node => node.tagName === "TABLE");
        if (!ioSampleTable) {
            throw new Error("Failed to find challenge output sample table.");
        }

        const initialCode = document
            .querySelector(`input[type="hidden"][name^="initial_code_"]`)
            ?.getAttribute("value");

        if (!initialCode) {
            throw new Error("Failed to parse initial code information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            ...parseSampleTable(ioSampleTable),
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            provider: this,
            initialCode,
        };
    }
}
