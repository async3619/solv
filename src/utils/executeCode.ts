import * as fs from "fs-extra";
import * as esbuild from "esbuild";
import { NodeVM } from "vm2";

import { BaseProvider } from "../providers/base";

import { Challenge, InputType } from "./types";

const replaceExt = require("replace-ext");

export async function executeCode(
    challenge: Challenge,
    input: InputType,
    output: string,
    targetPath: string,
    provider: BaseProvider,
    noTranspile = false,
): Promise<[string, string] | Error> {
    let fileContent = await fs.readFile(targetPath).then(res => res.toString());
    fileContent = provider.beforeExecute(fileContent, input, challenge);

    const transpiledContent = esbuild.transformSync(fileContent, {
        loader: "ts",
    });

    if (provider.needJs && !noTranspile) {
        const transpiledFilePath = replaceExt(targetPath, ".js");
        await fs.writeFile(transpiledFilePath, transpiledContent.code);
    }

    const outputBuffer: string[] = [];
    const debugBuffer: string[] = [];
    const handleConsoleMessage =
        (targetBuffer: string[]) =>
        (...data: any[]) => {
            targetBuffer.push(data.map(item => provider.serializeOutput(item)).join(" "));
        };

    const inputData = Array.isArray(input) ? input.map(item => JSON.parse(item)) : input;
    const vm = new NodeVM({
        console: "redirect",
        sandbox: {
            input: inputData,
        },
        require: {
            builtin: ["readline", "events"],
        },
        env: {
            arguments: inputData,
        },
    });

    vm.on("console.log", handleConsoleMessage(outputBuffer));
    vm.on("console.info", handleConsoleMessage(outputBuffer));
    vm.on("console.warn", handleConsoleMessage(outputBuffer));
    vm.on("console.error", handleConsoleMessage(outputBuffer));
    vm.on("console.debug", handleConsoleMessage(debugBuffer));

    try {
        vm.run(transpiledContent.code);
    } catch (e) {
        return e as Error;
    }

    return [outputBuffer.join("\n").replace(/\r\n/g, "\n"), debugBuffer.join("\n").replace(/\r\n/g, "\n")];
}
