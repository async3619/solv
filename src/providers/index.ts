import { BaseProvider } from "./base";

import { BaekjoonProvider } from "./baekjoon";
import { ProgrammersProvider } from "./programmers";
import { LeetCodeProvider } from "./leetcode";

const TARGET_PROVIDERS: BaseProvider[] = [new BaekjoonProvider(), new ProgrammersProvider(), new LeetCodeProvider()];

export function getProvider(targetUrl: string) {
    const provider = TARGET_PROVIDERS.find(provider => provider.checkUrl(targetUrl));
    if (!provider) {
        throw new Error("there was no matched coding challenge service provider on records. abort.");
    }

    return provider;
}
