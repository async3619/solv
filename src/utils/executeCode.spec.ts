import * as fs from "fs-extra";

import { BaekjoonProvider } from "../providers/baekjoon";
import { executeCode } from "./executeCode";

const SAMPLE_CODE = `
function solution(input: string[]) {
    console.log(input);
}

(callback => {
    if (typeof process !== "undefined" && "env" in process && "arguments" in process.env && process.env.arguments) {
        solution(process.env.arguments);
        return;
    }

    const readline = require("readline");
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const input: string[] = [];
    rl.on("line", line => {
        input.push(line);
    }).on("close", function () {
        callback(input);
        process.exit();
    });
})(solution);
`.trim();

describe("executeCode", () => {
    it("should transpiles and runs given code properly", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () =>
            Promise.resolve(`
function foo(input: string[]) {
    console.log("Hello World!");
    console.debug("Debug message.");
}

foo([]);
        `);
        (fs.writeFile as any) = () => Promise.resolve();

        const result = await executeCode({} as any, "Test", "Hello World!", "__MOCKED__", new BaekjoonProvider());
        expect(result).toStrictEqual(["Hello World!", "Debug message."]);
    });

    it("should pass given input data to code context", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(SAMPLE_CODE);
        (fs.writeFile as any) = () => Promise.resolve();

        expect(
            await executeCode({} as any, "Test", "Hello World!", "__MOCKED__", new BaekjoonProvider()),
        ).toMatchSnapshot();
    });

    it("should pass JSON serialized input data", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(SAMPLE_CODE);
        (fs.writeFile as any) = () => Promise.resolve();

        expect(
            await executeCode({} as any, [`["Test", "Test2"]`], "Hello World!", "__MOCKED__", new BaekjoonProvider()),
        ).toMatchSnapshot();
    });

    it("should catches error", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(`throw new Error("Test error")`);
        (fs.writeFile as any) = () => Promise.resolve();

        const result = await executeCode({} as any, "Test", "Hello World!", "__MOCKED__", new BaekjoonProvider());
        expect(result).toMatchSnapshot();
    });
});
