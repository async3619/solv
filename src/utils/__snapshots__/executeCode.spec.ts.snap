// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeCode should catches error 1`] = `[Error: Test error]`;

exports[`executeCode should pass JSON serialized input data 1`] = `
[
  "[["Test","Test2"]]",
  "",
]
`;

exports[`executeCode should pass given input data to code context 1`] = `
[
  "Test",
  "",
]
`;
